### PR TITLE
add AutoSemiDAIS

### PR DESCRIPTION
**Pull request by martinjankowiak**
@fehiepsi this is closer to working but there still appears to be some scaling issue.

run `pytest -s -v test_autoguide.py::test_autosemidais` to see apparent discrepancies.

### CODE REVIEW

File: numpyro/infer/autoguide.py
Comment (martinjankowiak) [inline numpyro/infer/autoguide.py:1027]: is (N / subsample_size) correct?

File: numpyro/infer/autoguide.py
Comment (martinjankowiak) [inline numpyro/infer/autoguide.py:1036]: does this need global_latents inside?

File: numpyro/infer/autoguide.py
Comment (martinjankowiak) [inline numpyro/infer/autoguide.py:1131]: is it true that local_log_density is scaled by (N / subsample_size) because of the plate?

File: numpyro/infer/autoguide.py
Comment (martinjankowiak) [inline numpyro/infer/autoguide.py:1153]: will this be scaled by (N / subsamplesize) because it's in the plate?

File: test/infer/test_autoguide.py
Comment (martinjankowiak) [inline test/infer/test_autoguide.py:725]: is there any reason reusing `svi_result16` like this would be incorrect?

**Review by fehiepsi — COMMENTED** [review]
@martinjankowiak I just removed the scaling factor of the grad term. It seems to work.

File: test/infer/test_autoguide.py
Comment (fehiepsi) [reply test/infer/test_autoguide.py:725]: I'm not sure. It seems to be valid.

**Comment by martinjankowiak** [comment]
> @martinjankowiak I just removed the scaling factor of the grad term. It seems to work.

this only appears to work because it's driven into a regime where eta ~= 0 and the hmc dynamics are trivial

```
    print(
        "svi_result16.params[auto_eta_coeff].mean()",
        svi_result16.params["auto_eta_coeff"].mean(),
        " (should be positive)",
    )
```

File: numpyro/infer/autoguide.py
Comment (martinjankowiak) [reply numpyro/infer/autoguide.py:1168]: if this is true we need the multiplicative factor in 

```
p_grad = beta[:, None] * (subsample_size / N) * grad(local_log_density)(z_half)
```
because `p_grad` and `q_grad` need to have the same scale. `q_grad` comes from `base_z_dist_log_prob` so i guess it should be unscaled

File: test/infer/test_autoguide.py
Comment (martinjankowiak) [reply test/infer/test_autoguide.py]: is it possible the parameters get mixed up somehow? so that the local parameters get assigned to the wrong datapoints? e.g. `z_0_loc`

**Comment by martinjankowiak** [comment]
@fehiepsi i think the mean thing to understand right now is why `dais_elbo8 != dais_elbo16`. these two code paths are supposed be computing the same elbos. the only difference should be the mini-batch size. in particular they should be equal in expectation and only differ by variance, which should be small for a large number of samples.

File: numpyro/infer/autoguide.py
Comment (fehiepsi) [reply numpyro/infer/autoguide.py:1168]: You  are right, just double check that we need such factor here.

**Comment by fehiepsi** [comment]
@martinjankowiak I just fixed a bug at unconstrained_reparam. Could you take a look? The elbos seem to be consistent now.

File: numpyro/infer/util.py
Comment (martinjankowiak) [inline numpyro/infer/util.py:221]: @fehiepsi can you please explain how exactly this logic effects our autoguide?

Comment (martinjankowiak) [reply numpyro/infer/util.py:221]: i'd like to understand better where exactly this bug would be triggered. only inside a plate? both factor and sample? etc. in particular i'd like to understand which different code path in local.py avoided this issue...

Comment (martinjankowiak) [reply numpyro/infer/util.py:221]: in any case the test now behaves as expected : ) thanks! 💯

File: numpyro/distributions/distribution.py
Comment (martinjankowiak) [inline numpyro/distributions/distribution.py:561]: we still need to double check this...

File: numpyro/infer/autoguide.py
Comment (martinjankowiak) [inline numpyro/infer/autoguide.py:910]: are these the right/complete requirements for local_model and base_guide?

File: numpyro/infer/autoguide.py
Comment (martinjankowiak) [inline numpyro/infer/autoguide.py:941]: this is only used to setup `_local_potential_fn_gen` right? i.e. the value of `init_loc_fn` doesn't matter?

File: numpyro/infer/autoguide.py
Comment (martinjankowiak) [inline numpyro/infer/autoguide.py:989]: is this ok?

File: numpyro/infer/autoguide.py
Comment (martinjankowiak) [inline numpyro/infer/autoguide.py:1229]: still need to implement this

**Comment by martinjankowiak** [comment]
@fehiepsi ok i think this should be good to go once we resolve the remaining comments/todos above.

any additional tests we should add?

File: numpyro/infer/util.py
Comment (fehiepsi) [reply numpyro/infer/util.py:221]: As in the new tests that I added, when we do unconstrained reparam, we introduce new factor terms. Those factor terms will belong to the same context as the corresponding latent variables, i.e. they will have the same `scale` field in their messages. So with those lines, we are scaling logdet terms 2 times.

Comment (martinjankowiak) [reply numpyro/infer/util.py:221]: ah gotcha i think that makes sense

File: numpyro/distributions/distribution.py
Comment (fehiepsi) [reply numpyro/distributions/distribution.py:561]: I think we can add `intermediates` arguments to log_prob method to utilize this information. This is useful for Expanded(TransformDistribution).

File: numpyro/handlers.py
Comment (fehiepsi) [reply numpyro/handlers.py:262]: I think we can revert this change because we don't need this. Or keep this as-is if you think expose_types is useful for something else

File: numpyro/infer/autoguide.py
Comment (fehiepsi) [reply numpyro/infer/autoguide.py:910]: I think so. Which information else that you have in mind?

File: numpyro/distributions/distribution.py
Comment (martinjankowiak) [reply numpyro/distributions/distribution.py:561]: ok i'll leave this to you 👍

File: numpyro/infer/autoguide.py
Comment (martinjankowiak) [reply numpyro/infer/autoguide.py:1078]: is it enough to add `result.update(global_latents)`?

**Review by fehiepsi — APPROVED** [review]
LGTM!

File: numpyro/infer/autoguide.py
Comment (fehiepsi) [reply numpyro/infer/autoguide.py]: that's right

**Review by martinjankowiak — COMMENTED** [review]
@fehiepsi great thanks for all your help!

**Pull request merged**
